src/libstd/collections/hash/map.rs

```diff
@@ -1370,6 +1370,37 @@ impl<K, V, S> Eq for HashMap<K, V, S>
 {
 }
 
+#[unstable(feature = "hashmap_hash", issue = "0")]
+impl<K, V, S> Hash for HashMap<K, V, S>
+    where K: Eq + Hash,
+          V: Hash,
+          S: BuildHasher
+{
+    fn hash<H: Hasher>(&self, hasher: &mut H) {
+        // We must preserve:  x == y -> hash(x) == hash(y).
+        // HashMaps have no order, so we must use a commutative operation
```

Comment:
To be precise, the operation must be associative as well, because in theory you're iterating over arbitrary permutations of the elements. Fortunately, `wrapping_add` is both.

Reply:
I'll note that. =)